docs/javascript-unit-tests.md

```diff
@@ -0,0 +1,712 @@
+            image: 'https://vignette.wikia.nocookie.net/tmnt/images/0/00/Krangnobody.png',
+            caption: 'Krang portrait'
+        };
+        expect( sample.gimmeObject() ).toBe( sampleObject );
+      } );
+
+    } );
+
+} );
+```
+
+In these tests, we check _(`expect`)_ that
+the string returned by `sample.gimmeString()`
+and the object returned by `sample.gimmeObject()`
+match _(`toBe`)_ the `sampleString` and `sampleObject`
+that we defined in the tests.
+
+There are many kinds of assertions you can use
+besides the simple equality comparison of `toBe`.
+See [the Jest Using Matchers guide](https://jestjs.io/docs/en/using-matchers)
+for a primer on them and
+[the Jest `expect` API docs](https://jestjs.io/docs/en/expect)
+for the full list of its matchers.
+
+
+### Providing test data
+
+The first principle of test data for unit tests is to
+**keep test data as simple as possible** –
+use the minimum needed to test the code.
+
+#### Direct definition of test data
+
+The simplest way to set up test data is to
+declare it as variables within each test, e.g., the tests in
+[strings-spec.js](https://github.com/cfpb/cfgov-refresh/blob/master/test/unit_tests/js/modules/util/strings-spec.js).
+This can include HTML markup for DOM manipulation tests,
+if each test requires different markup.
+
+#### Setup and teardown methods
+
+If you will need to leverage the same test data across different tests,
+Jest has setup and teardown methods, such as
+[`beforeEach` and `afterEach`](https://jestjs.io/docs/en/setup-teardown#repeating-setup-for-many-tests), or [`beforeAll` and `afterAll`](https://jestjs.io/docs/en/setup-teardown#one-time-setup),
+which can be used to performing actions
+that are needed before and after running all tests or each test in a suite.
+For example, the tests in
+[`Analytics-spec.js`](https://github.com/cfpb/cfgov-refresh/blob/add-js-unit-testing-doc/test/unit_tests/js/modules/Analytics-spec.js#L7-L32)
+use both `beforeAll` and `beforeEach` inside the root `describe` block to
+do a variable definition for all tests at the beginning of the suite
+and reset the `dataLayer` before each test, respectively.
+
+[Check out the Jest documentation on “Setup and teardown” methods.](https://jestjs.io/docs/en/setup-teardown)
+
+A common structure when the DOM is involved
+is to create a constant representing an HTML snippet to test,
+then – in a `beforeEach` or `beforeAll`
+(depending on whether the tests modify the markup or not) –
+set `document.body.innerHTML` to that snippet.
+
+Use `beforeAll` to attach HTML markup that is unaffected by the tests,
+e.g., the tests in
+[`footer-button-spec.js`](https://github.com/cfpb/cfgov-refresh/blob/add-js-unit-testing-doc/test/unit_tests/js/modules/footer-button-spec.js).
+Use `beforeEach` to reset manipulated markup between tests, e.g., the tests in
+[`Notification-spec.js`](https://github.com/cfpb/cfgov-refresh/blob/add-js-unit-testing-doc/test/unit_tests/js/molecules/Notification-spec.js).
```

Comment:
This links to this PR branch, which might be deleted in future. Link to a specific commit SHA instead?